hivemind/optim/experimental/optimizer.py

```diff
@@ -0,0 +1,460 @@
+    Parameters,
+    ParamGroups,
+    SchedulerFactory,
+    TorchOptimizer,
+    TrainingStateAverager,
+)
+from hivemind.optim.grad_scaler import GradScaler
+from hivemind.utils import get_dht_time, get_logger
+
+logger = get_logger(__name__)
+
+
+class Optimizer(torch.optim.Optimizer):
+    """
+    Hivemind Optimizer wraps your regular PyTorch Optimizer for training in a swarm of peers. It can be configured with
+     synchronous, delayed or asynchronous updates to trade between optimization guarantees and compute utilization.
+
+    The Optimizer is meant as a drop-in replacement for your regular PyTorch code:
+
+    >>> model = transformers.AutoModel("albert-xxlarge-v2")
+    >>> dht = hivemind.DHT(initial_peers=INITIAL_PEERS, start=True)
+    >>> opt = hivemind.Optimizer(model.parameters(), optim_cls=torch.optim.Adam, prefix="run_42",
+    >>>                          target_batch_size=4096, batch_size_per_step=4)
+    >>> while True:
+    >>>     loss = compute_loss_on_batch(model, batch_size=4)
+    >>>     opt.zero_grad()
+    >>>     loss.backward()
+    >>>     opt.step()  # <-- train collaboratively with any peers that use the same prefix (run_42)
+
+    However, unlike regular optimizers, calling opt.step with hivemind.Optimizer can do one of the following:
+    - accumulate a minibatch of data towards the (global) target batch size without changing parameters (yet),
+    - after accumulating the target batch size, all-reduce gradients with peers and perform optimizer step,
+    - if, for any reason, your peer lags behind the rest of the swarm, it will load state from up-to-date peers.
+
+    :note: Hivemind.Optimizer can be used the same way any other pytorch optimizer, but there is one limitation:
+      learning rate schedulers, curriculum and other time-dependent features should use opt.global_step (and not the
+      number of local forward-backward cycles). This is because any device can join midway through training, when
+      other peers have already made some progress and changed their learning rate accordingly.
+
+    :param dht: a running hivemind.DHT instance connected to other peers
+    :param prefix: a unique name of this experiment, used as a common prefix for all DHT keys
+    :param target_batch_size: perform optimizer step after all peers collectively accumulate this many samples
+    :param batch_size_per_step: before each call to .step, user should accumulate gradients over this many samples
+    :param optimizer: a standard pytorch optimizer, preferably a large-batch one such as LAMB, LARS, etc.
+    :param params: optional, a list/tuple of parameters or structured param groups for the optimizer
+    :param scheduler: if specified, use this scheduler to update optimizer learning rate
+    :note: If you are using ColloptaborativeOptimizer with lr_scheduler, it is recommended to pass this scheduler
+      explicitly into this class. Otherwise, scheduler may not be synchronized between peers.
```

Comment:
```suggestion
    :note: If you are using hivemind.Optimizer with lr_scheduler, it is recommended to pass this scheduler
      explicitly into this class. Otherwise, it may become non-synchronized between peers.
```